pkg/clusterapi/etcd.go

```diff
@@ -0,0 +1,86 @@
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+// SetEtcdConfigInEtcdCluster sets up the etcd config in EtcdadmCluster.
+func SetEtcdConfigInEtcdCluster(etcd *etcdv1.EtcdadmCluster, version string) {
+	etcd.Spec.EtcdadmConfigSpec.CloudInitConfig = &etcdbootstrapv1.CloudInitConfig{
+		Version:    version,
+		InstallDir: "/usr/bin",
+	}
+	etcd.Spec.EtcdadmConfigSpec.PreEtcdadmCommands = []string{
+		"hostname \"{{`{{ ds.meta_data.hostname }}`}}",
+		"echo \"::1         ipv6-localhost ipv6-loopback\" >/etc/hosts",
+		"echo \"127.0.0.1   localhost\" >>/etc/hosts",
+		"echo \"127.0.0.1   {{`{{ ds.meta_data.hostname }}`}}\" >>/etc/hosts",
+		"echo \"{{`{{ ds.meta_data.hostname }}`}}\" >/etc/hostname",
+	}
+}
+
+// SetEtcdConfigInCluster sets up the etcd config in CAPI Cluster.
+func SetEtcdConfigInCluster(cluster *clusterv1.Cluster, clusterSpec *cluster.Spec) {
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration == nil {
+		return
+	}
+
+	cluster.Spec.ManagedExternalEtcdRef = &v1.ObjectReference{
+		APIVersion: etcdClusterAPIVersion,
+		Kind:       etcdadmClusterKind,
+		Name:       EtcdClusterName(clusterSpec),
+		Namespace:  constants.EksaSystemNamespace,
+	}
+}
+
+// SetUnstackedEtcdConfigInKubeadmControlPlaneForBottlerocket sets up unstacked etcd configuration in kubeadmControlPlane for bottlerocket.
+func SetUnstackedEtcdConfigInKubeadmControlPlaneForBottlerocket(kcp *controlplanev1.KubeadmControlPlane, externalEtcdConfig *v1alpha1.ExternalEtcdConfiguration) {
+	if externalEtcdConfig == nil {
+		return
+	}
+
+	kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External = &bootstrapv1.ExternalEtcd{
+		Endpoints: []string{},
+		CAFile:    "/var/lib/kubeadm/pki/etcd/ca.crt",
+		CertFile:  "/var/lib/kubeadm/pki/server-etcd-client.crt",
+		KeyFile:   "/var/lib/kubeadm/pki/apiserver-etcd-client.key",
+	}
+}
+
+// SetUnstackedEtcdConfigInKubeadmControlPlaneForUbuntu sets up unstacked etcd configuration in kubeadmControlPlane for ubuntu.
+func SetUnstackedEtcdConfigInKubeadmControlPlaneForUbuntu(kcp *controlplanev1.KubeadmControlPlane, externalEtcdConfig *v1alpha1.ExternalEtcdConfiguration) {
+	if externalEtcdConfig == nil {
+		return
+	}
+
+	kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External = &bootstrapv1.ExternalEtcd{
+		Endpoints: []string{},
+		CAFile:    "/etc/kubernetes/pki/etcd/ca.crt",
+		CertFile:  "/etc/kubernetes/pki/apiserver-etcd-client.crt",
+		KeyFile:   "/etc/kubernetes/pki/apiserver-etcd-client.key",
+	}
+}
+
+// SetStackedEtcdConfigInKubeadmControlPlane sets up stacked etcd configuration in kubeadmControlPlane.
+func SetStackedEtcdConfigInKubeadmControlPlane(kcp *controlplanev1.KubeadmControlPlane, clusterSpec *cluster.Spec) {
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
```

Comment:
yeah, I feel like all these checks should probably be in the caller

Reply:
why? does it make big difference?